---
title: IO异常处理
date: 2020-10-23 17:47:33
tags:	
	- Java
	- IO
	- 基础知识
categories:
	- Java
cover:
	https://gitee.com/jasonM4A1/pictureHost/raw/master/20201023174831.jpg
---

# IO异常处理

## JDK7前处理方法

之前的入门练习，我们一直把异常抛出，而实际开发中并不能这样处理，在`JDK 7`之前我们都是建议使用`try...catch...finally` 代码块处理异常部分。

**代码演示**

~~~java
import java.io.IOException;
import java.io.FileOutputStream;
import java.io.FileInputStream;

/**
 * 完成图片的copy，
 * 同时使用try...catch..finally处理异常部分
 */
public class Test {
    public static void main(String[] args) {
        FileInputStream fis = null;
        FileOutputStream fos = null;
        try {
            //读取的文件流对象
            fis = new FileInputStream("D:\\pictures\\wallhaven-zm1dzy.jpg");
            //写出的文件流对象
            fos = new FileOutputStream("D:\\study\\Java\\node\\heima\\wallhaven-zm1dzy_copy.jpg");

            //接收读取到的字节
            byte[] bytes = new byte[1024];
            //接收读取到的有效字节个数
            int b = 0;

            //进行图片的copy
            while ((b = fis.read(bytes)) != -1) {
                fos.write(bytes, 0, b);
            }
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            //关闭写出的文件流对象
            try {
                if (fos != null) {
                    fos.close();
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
            //关闭读取的文件流对象
            try {
                if (fis != null) {
                    fis.close();
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }
}
~~~

## JDK 7的优化处理

到了`JDK7`，Java对异常的处理进行了优化，可以使用`try-with-resource` 语句了，该语句确保了每个资源在语句结束时关闭。所谓的资源（resource）是指在程序完成后，必须关闭的对象。

### 格式

~~~java
try (创建流对象语句，多个流对象，使用;隔开) {
    //对流对象进行操作的代码
} catch (IOException e) {
	//一些处理异常的代码，如e.printStackTrace();
}
~~~

> 采用`try-with-resource`语句，就不用每次在`finally`代码块中手动写入关闭流对象的代码了。在`try ()`中创建的流对象，在使用完毕后，JVM会自动关闭这些流对象。

**代码演示**

~~~java
import java.io.IOException;
import java.io.FileReader;
import java.io.FileWriter;

/**
 * 完成文本文件的copy
 * 同时使用try-with-resource语句，对IO异常进行处理
 */
public class Test {
    public static void main(String[] args) {
        //在try后的小括号里创建多个流对象，使用后会被JVM自动关闭
        try (FileReader fr = new FileReader("D:\\study\\Java\\node\\heima\\29-Map集合.md");
             FileWriter fw = new FileWriter("D:\\pictures\\29-Map集合_copy.md")) {
            //对文本文件进行copy
            int b = 0;
            char[] chars = new char[2];
            while ((b = fr.read(chars)) != -1) {
                fw.write(chars);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
~~~

## JDK 9的改进

JDK9中`try-with-resource` 的改进，对于**引入对象**的方式，支持的更加简洁。被引入的对象，同样可以自动关闭，无需手动close，我们来了解一下格式。

**改进前格式：**

~~~java
//被final修饰的对象
final Resource resource1 = new Resource("resource1");
//普通对象
Resource resource2 = new Resource("resource2");
//引入方式：创建新的变量保存
try (Resource r1 = resource1; Resource r2 = resource2) {
    //使用上面创建的对象
} catch (异常类型 变量) {
    //对异常处理的代码
}
~~~

**改进后格式：**

~~~java
//被final修饰的对象
final Resource resource1 = new Resource("resource1");
//普通对象
Resource resource2 = new Resource("resource2");
//引入方法：直接引入
try (resource1; resource2) {
    //使用上面创建的对象
} catch (异常类型 变量) {
	//对异常处理的代码
}
~~~

**改进后，代码演示：**

~~~java
import java.io.IOException;
import java.io.FileReader;
import java.io.FileWriter;

/**
 * 完成文本文件的copy
 * 同时使用try-with-resource JDK-9改进的语句，对IO异常进行处理
 */
public class Test {
    public static void main(String[] args) throws IOException {
        //创建流对象
        FileReader fr = new FileReader("D:\\study\\Java\\node\\heima\\29-Map集合.md");
        FileWriter fw = new FileWriter("D:\\pictures\\29-Map集合_copy.md");
        //引入到try中
        try (fr; fw) {
            int b = 0;
            char[] chars = new char[2];
            while ((b = fr.read(chars)) != -1) {
                fw.write(chars);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
~~~

